### ISSUE TITLE
supporting custom resolvers per cluster

### ISSUE THREAD

**Issue by tomwans** [lyft/envoy#940]
My team has a need for Envoy to support custom DNS resolvers per cluster (we need different `strict_dns` clusters to query specific resolvers). We have a small PoC working internally, but to do it right would involve quite a bit of surgery. Namely, we'd need to create a Network::DnsResolver per cluster vs. creating it up-front in the server and passing that one down to all the cluster managers and clusters.

Is the envoy team interested in such a patch? We would be happy to contribute it, but if it's not something y'all are looking for, we'd like to know so we can just keep it internal.

**Comment by mattklein123** [comment]
This is not a feature request that I've heard before. With that said, if you need it in production, it sounds like something we should explore.

I think we can probably simplify by not changing the global resolver, but instead just allow a custom resolver to be optionally created and used on a per cluster basis. This should be a lot less surgery. I assume fundamentally you need to specify a resolver address in the config?

If you want to contribute can you post a summary of the proposed config changes, as well as a short design sketch of what you would change and we can go from there?

**Comment by tomwans** [comment]
> I think we can probably simplify by not changing the global resolver, but instead just allow a custom resolver to be optionally created and used on a per cluster basis.

Yep, that's the plan. What's the need for the global resolver? I was able to comment it out (dns_resolver_) from the InstanceImpl without an issue. Maybe I'm not understanding exactly what the global resolver is?

> I assume fundamentally you need to specify a resolver address in the config?

Yep, as far as the configuration changes go, I think we'd have a new entry inside of the cluster type, similar to "dns_refresh_rate_ms", called "dns_resolvers". It would take an array of resolver IP:Port strings to set as the resolvers. Taking in IP:port strings is already built in to c-ares, so that's easy. If it's empty, or unspecified, we would configure the DnsResolver as normal, implicitly reading from resolv.conf.

I don't have time to get into the design sketch today, but I can get to that soon.

**Comment by htuch** [comment]
@mattklein123 We're also thinking it might be useful to have more flexible resolver configuration. Specifically, the ability to transform arbitrary `Address` in the v2 API via a plugin for both listeners and upstream addresses. The plugin might use DNS or something else. This goes beyond what @tomwans is talking about, but just something to keep in mind. @tschroed.

**Comment by mattklein123** [comment]
@htuch OK let's sync offline a bit. Not totally clear what you mean.

@tomwans the only reason to keep the global resolver around is that in the general case, it will use less memory and connections for many DNS clusters. I think it's probably worth it to keep it, but could be convinced otherwise. Also, qq, are we sure that c-ares doesn't have an API that would allow us to specify resolver details on a per resolve basis? (I haven't looked). If it does, another option is to still use a global resolver but have a version of resolve() that allows specification of the resolver address.

**Comment by htuch** [comment]
c-ares does allow you to set the DNS servers on a per-ares_channel basis (https://c-ares.haxx.se/ares_set_servers_ports_csv.html), we have one channel per resolver today.

**Comment by htuch** [comment]
https://c-ares.haxx.se/ares_set_servers.html as well

**Comment by tomwans** [comment]
Yep, it seems like c-ares only allows you to set the resolvers per-channel; that's what I've done in our internal fork.

Here's the diff from our internal fork, and provides an overview of the general change I'm proposing: https://gist.github.com/tomwans/ef569b9fc070c9b9b56c797a679abacd. 

TL;DR most of the work is done inside of `ClusterImplBase::create` (https://gist.github.com/tomwans/ef569b9fc070c9b9b56c797a679abacd#file-envoy-diff-L202), where we conditionally create a new `DnsResolver` and pass it into the `StrictDnsClusterImpl`. `StrictDnsClusterImpl` also gets a funky class member called `custom_dns_resolver_` which owns the custom resolver in a `DnsResolverPtr`. I'll take any guidance on what to do with that, I don't know if that's considered awkward form or not. (I'm not super experienced with C++)

We could do the same thing to the logical_dns cluster type.

**Comment by mattklein123** [comment]
@tomwans there are a few issues in the diff, but at a high level it looks fine to me. Feel free to do a PR and we can cover the issues during the normal code review process.

**Comment by tomwans** [comment]
@mattklein123 word, will do next week

**Comment by mattklein123** [comment]
Fixed.

**Comment by tomwans** [comment]
Thanks for all ya'lls help. Feels good to not have to run a fork anymore :)

Feel free to ping me with any questions or if issues arise from this change, I'm happy to help!
